### PR TITLE
cmake,rgw: make amqp support optional

### PR DESCRIPTION
**Pull request by tchaikov**
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -440,6 +435,7 @@ option(WITH_RADOSGW "Rados Gateway is enabled" ON)
 option(WITH_RADOSGW_FCGI_FRONTEND "Rados Gateway's FCGI frontend is enabled" OFF)
 option(WITH_RADOSGW_BEAST_FRONTEND "Rados Gateway's Beast frontend is enabled" ON)
 option(WITH_RADOSGW_BEAST_OPENSSL "Rados Gateway's Beast frontend uses OpenSSL" ON)
+option(WITH_RADOSGW_AMQP_ENDPOINT "Rados Gateway's pubsub support for AMQP push endpoint" ON)
```
Comment (jecluis) [inline CMakeLists.txt]: not very familiar with cmake, tbh, but is this checking the same things as the code you removed above? I see the above code checking for the header, but this seems to only define the option, without specifying the required header.

Comment (jecluis) [reply CMakeLists.txt]: well, disregard. That's happening below, in the `FindRabbitMQ.cmake` file.

File: src/rgw/rgw_pubsub_push.cc [hunk 2]
```diff
@@ -314,6 +320,7 @@ RGWPubSubEndpoint::Ptr RGWPubSubEndpoint::create(const std::string& endpoint,
       throw configuration_error("unknown amqp version " + version);
       return nullptr;
     }
+#endif
```
Comment (yuvalif) [inline src/rgw/rgw_pubsub_push.cc]: probably better to put the amqps case into the ifdef

File: install-deps.sh [hunk 3]
```diff
@@ -272,7 +272,6 @@ else
         echo "Using apt-get to install dependencies"
         $SUDO apt-get install -y devscripts equivs
         $SUDO apt-get install -y dpkg-dev
```
Comment (yuvalif) [inline install-deps.sh]: see line 253 - do we want to make it optional in freebsd as well?

**Comment by yuvalif** [comment]
The AMQP unit-test should be removed from the rgw unit-test list.
Also, no need to build the mock AMQP library in this case.

**Comment by yuvalif** [comment]
@jecluis is it possible to add opensuse build to CI, so this such issues are detectable before a PR is merged?

File: install-deps.sh (hunk 3 above)
Comment (tchaikov) [reply install-deps.sh]: i think it's up to @wjwithagen

Comment (wjwithagen) [reply install-deps.sh]: Yes, please.....
I'm very much in favour that all nuts and bolts glued on, are conditionalised.
Especially when they are not essential and required for the functioning of the main Ceph components.
And what I like even beter is that with this PR is that in absence of `rabbitmq` this code is now just skipped, but RGW is still build.
Also I have no way of verifying that this will work on FreeBSD, since I do not have the infrastructure that goes with it.

So in short I would not put it in the `pkg install` list for FreeBSD.

**Comment by tchaikov** [comment]
@idryomov the issue should have been fixed in the latest change. could you try again?

File: src/rgw/rgw_pubsub_push.cc (hunk 2 above)
Comment (tchaikov) [reply src/rgw/rgw_pubsub_push.cc]: done.

File: install-deps.sh (hunk 3 above)
Comment (tchaikov) [reply install-deps.sh]: added a commit to disable amqp endpoint for FreeBSD.

Comment (yuvalif) [reply install-deps.sh]: why not make it conditional?
(BTW building was verified on freebsd12)

Comment (tchaikov) [reply install-deps.sh]: @yuvalif what do you mean by conditional?

Comment (yuvalif) [reply install-deps.sh]: @tchaikov i mean, based on the build flag, similar to how we do that for other OSs

Comment (tchaikov) [reply install-deps.sh]: @yuvalif the cmake script handles various `WITH_*` already. on rpm based distro we have with and without conditions. on debian derivatives, i don't think we have any "build flag" aside from cmake options. probably you could help send a follow-up PR to address this? and i am happy to drop the commit disabling this feature if it makes more sense to you and @wjwithagen ?

**Comment by liewegas** [comment]
could tehse build errors be related?
https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos7,DIST=centos7,MACHINE_SIZE=huge/18135//consoleFull
```
CMakeFiles/unittest_rgw_dmclock_scheduler.dir/test_rgw_dmclock_scheduler.cc.o: In function `auto rgw::dmclock::AsyncScheduler::async_request<boost::asio::executor_binder<rgw::dmclock::capture(std::optional<boost::system::error_code>&, std::optional<crimson::dmclock::PhaseType>&)::{lambda(boost::system::error_code, crimson::dmclock::PhaseType)#1}, boost::asio::io_context::executor_type> >(rgw::dmclock::client_id const&, crimson::dmclock::ReqParams const&, double const&, unsigned int, boost::asio::executor_binder<rgw::dmclock::capture(std::optional<boost::system::error_code>&, std::optional<crimson::dmclock::PhaseType>&)::{lambda(boost::system::error_code, crimson::dmclock::PhaseType)#1}, boost::asio::io_context::executor_type>&&) [clone .constprop.997]':
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/14.0.1-4068-g79f82a4/rpm/el7/BUILD/ceph-14.0.1-4068-g79f82a4/src/rgw/rgw_dmclock_async_scheduler.h:153: undefined reference to `PerfCounters::inc(int, unsigned long)'
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/14.0.1-4068-g79f82a4/rpm/el7/BUILD/ceph-14.0.1-4068-g79f82a4/src/rgw/rgw_dmclock_async_scheduler.h:154: undefined reference to `PerfCounters::inc(int, unsigned long)'
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/14.0.1-4068-g79f82a4/rpm/el7/BUILD/ceph-14.0.1-4068-g79f82a4/src/rgw/rgw_dmclock_async_scheduler.h:140: undefined reference to `rgw::dmclock::AsyncScheduler::schedule(double const&)'
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/14.0.1-4068-g79f82a4/rpm/el7/BUILD/ceph-14.0.1-4068-g79f82a4/src/rgw/rgw_dmclock_async_scheduler.h:142: undefined reference to `PerfCounters::inc(int, unsigned long)'
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/14.0.1-4068-g79f82a4/rpm/el7/BUILD/ceph-14.0.1-4068-g79f82a4/src/rgw/rgw_dmclock_async_scheduler.h:143: undefined reference to `PerfCounters::inc(int, unsigned long)'
CMakeFiles/unittest_rgw_dmclock_scheduler.dir/test_rgw_dmclock_scheduler.cc.o: In function `auto rgw::dmclock::AsyncScheduler::async_request<rgw::dmclock::capture(std::optional<boost::system::error_code>&, std::optional<crimson::dmclock::PhaseType>&)::{lambda(boost::system::error_code, crimson::dmclock::PhaseType)#1}>(rgw::dmclock::client_id const&, crimson::dmclock::ReqParams const&, double const&, unsigned int, rgw::dmclock::capture(std::optional<boost::system::error_code>&, std::optional<crimson::dmclock::PhaseType>&)::{lambda(boost::system::error_code, crimson::dmclock::PhaseType)#1}&&) [clone .constprop.1006]':
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/14.0.1-4068-g79f82a4/rpm/el7/BUILD/ceph-14.0.1-4068-g79f82a4/src/rgw/rgw_dmclock_async_scheduler.h:153: undefined reference to `PerfCounters::inc(int, unsigned long)'
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/14.0.1-4068-g79f82a4/rpm/el7/BUILD/ceph-14.0.1-4068-g79f82a4/src/rgw/rgw_dmclock_async_scheduler.h:154: undefined reference to `PerfCounters::inc(int, unsigned long)'
...
```

File: src/test/CMakeLists.txt [hunk 4]
```diff
@@ -53,6 +53,10 @@ add_subdirectory(osdc)
 add_subdirectory(pybind)
 if(${WITH_RADOSGW})
   add_subdirectory(rgw)
+  set(rgw_libs rgw_a)
+  if(WITH_RADOSGW_AMQP_ENDPOINT)
+    list(APPEND rgw_libs amqp_mock)
```
Comment (cbodley) [inline src/test/CMakeLists.txt]: `${rgw_libs}` is used in test/rgw/CMakeLists.txt, but this variable is set after `add_subdirectory(rgw)`

Comment (tchaikov) [reply src/test/CMakeLists.txt]: fixed and repushed.

**Comment by tchaikov** [comment]
https://shaman.ceph.com/builds/ceph/wip-optional-amqp/818e1031d03695c2479fb4b540bcc2f287a15175/
